File: trabalhoFinal/src/trabalho/Cliente.java


package trabalho;

import java.util.ArrayList;
import java.util.List;
import java.time.LocalDateTime;

public class Cliente extends Pessoa {
    private int idCliente;
    private String email;
    protected List<Venda> historicoCompras;

    public Cliente(int idCliente, String nome, String cpf, String telefone, String endereco, String email) {
        super(nome, cpf, telefone, endereco);
        this.idCliente = idCliente;
        this.email = email;
        this.historicoCompras = new ArrayList<>();
    }

    public int getIdCliente() {
        return idCliente;
    }

    public void setIdCliente(int idCliente) {
        this.idCliente = idCliente;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public List<Venda> getHistoricoCompras() {
        return historicoCompras;
    }

    public void adicionarCompraAoHistorico(Venda venda) {
        if (venda != null) {
            this.historicoCompras.add(venda);
        }
    }

    @Override
    public void consultarHistorico() {
        System.out.println("--- Histórico de Compras do Cliente: " + this.getNome() + " ---");
        System.out.println("ID Cliente: " + this.idCliente + " | CPF: " + this.getCpf());

        if (historicoCompras.isEmpty()) {
            System.out.println("\nNenhuma compra registrada.");
            return;
        }

        for (Venda venda : historicoCompras) {
            System.out.println(String.format(
                "\n  - Venda ID: %d | Status: %s | Valor Total: R$%.2f",
                venda.getIdVenda(),        
                venda.getStatusVenda(),    
                venda.getValorTotal()      
            ));

            System.out.println("    Itens Comprados:");

            List<Produto> produtosDaVenda = venda.getProdutos();
            List<Integer> quantidadesDaVenda = venda.getQuantidades();
            
            for (int i = 0; i < produtosDaVenda.size(); i++) {
                Produto produto = produtosDaVenda.get(i);
                Integer quantidade = quantidadesDaVenda.get(i);
                
                System.out.println(String.format(
                    "      - %s (%dx)",
                    produto.getNome(), 
                    quantidade
                ));
            }
        }
        System.out.println("\n" + "-".repeat(50));
    }
    

    @Override
    public String toString() {
        return
                " Cliente: " + idCliente +
                ", Pessoa: " + super.toString() + // Inclui nome, cpf, telefone, endereco da classe Pessoa
                ", Email: " + email;
    }
}

File: Python/cliente.py

from pessoa import Pessoa 

class Cliente(Pessoa):
    """Representa um cliente e gerencia seu histórico de compras."""
    def __init__(self, nome: str, cpf: str, telefone: str, endereco: str, id_cliente: int):
        super().__init__(nome, cpf, telefone, endereco)
        self._id_cliente = id_cliente
        self.historico_compras = [] # A lista começa vazia

    @property
    def id_cliente(self) -> int:
        return self._id_cliente
    
    def adicionar_venda_ao_historico(self, venda):
        self.historico_compras.append(venda)

    def consultar_historico(self):
        print(f"\n--- Histórico de Compras de: {self.nome} ---")
        if not self.historico_compras:
            print("Nenhuma compra registrada.")
            return

        for venda in self.historico_compras:
            print(f"\n  ID da Venda: {venda._id_venda} | Status: {venda._status} | Total: R$ {venda.valor_total:.2f}")
            for produto, quantidade in zip(venda._lista_de_produtos, venda._lista_de_quantidades):
                print(f"    - Item: {produto.nome} ({quantidade}x)")
        
        print("-" * 40)

    def __str__(self):
        return f"Cliente [ID: {self.id_cliente}] - {super().__str__()}"